### PR TITLE
Native Support for Typings

### PR DESCRIPTION
**Pull request by christian-bromann**
_From @christian-bromann on July 10, 2018 12:31_

## The problem

Many users are using WebdriverIO with typescript. We should ensure that these people have a first level support for the new object structure and commands.

Support should be added for:
- webdriver
- webdriverio
- any service that adds commands

_Copied from original issue: webdriverio/v5#86_

### CODE REVIEW

**Comment by erwinheitzman** [comment]
Code looks alright to me. Also I have been extremely busy and was struggling to implement this with the knowledge that I have about how things are set up.
Looking at the code makes it seems so simple afterwards though hehe. Really like the implementation.

**Comment by christian-bromann** [comment]
Anyone feel free to take over.

**Comment by timbru31** [comment]
Sorry for my late feedback, I'm going to take a look at this in the next days.

**Comment by christian-bromann** [comment]
@timbru31 it is not quite ready yet, I will post a gist with the generated type definitions soon

**Comment by timbru31** [comment]
Do you plan to maintain them in the webdriverio project itself or publish them via @types?

**Comment by christian-bromann** [comment]
> Do you plan to maintain them in the webdriverio project itself or publish them via @types?

I plan to maintain them here, what would you suggest tho?

**Comment by dylang** [comment]
IMHO: In this project only. `@types` is for when project owners don't include types.

**Comment by dylang** [comment]
Might want to include as a devDependency something like [tsd-check](https://github.com/SamVerschueren/tsd-check) to validate the generated types.

**Comment by dylang** [comment]
From Sindre Sorhus: [typescript-definition-style-guide](https://github.com/sindresorhus/typescript-definition-style-guide). It includes instructions for using `tsd-check`.

**Comment by ChrisOelmueller** [comment]
You probably wanted @christian-bromann ;-)

**Comment by christian-bromann** [comment]
@Bamieh you can continue working on the existing branch and finish the typing generation

**Comment by christian-bromann** [comment]
Taking it out for the v5 release milestone.

File: scripts/generate-typings.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+const lines = []
+for (const [protocolName, definition] of Object.entries(PROTOCOLS)) {
+    lines.push(`    // ${protocolName} types`)
+    lines.push('    interface Client<T> {')
+
+    for (const [, methods] of Object.entries(definition)) {
+        for (const [, description] of Object.entries(methods)) {
+            const { command, parameters = [], returns } = description
+            const params = parameters.map((p) => `${p.name}: ${p.type.toLowerCase()}`)
+            const returnValue = returns ? returns.type.toLowerCase() : 'undefined'
```
Comment (StephenABoyd) [inline scripts/generate-typings.js]: Some of the return types in the JSON files are of type ```*```. These should be manipulated here to ```any``` or changed in the JSON

File: scripts/generate-typings.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            lines.push(`        ${command}(${params.join(', ')}): ${returnValue}`)
```
Comment (StephenABoyd) [inline scripts/generate-typings.js]: While not 100% necessary, these lines should end with a semicolon

Comment (StephenABoyd) [reply scripts/generate-typings.js]: This should also pull the ```required``` option, and if the parameter is not required then it should be set as optional

**Comment by StephenABoyd** [comment]
@christian-bromann Do you know what should be in the ```Timeout``` interface?

**Comment by christian-bromann** [comment]
@StephenABoyd what timeout interface?

**Comment by StephenABoyd** [comment]
@christian-bromann The ```Timeouts``` interface or type. I figured it was just ```script```,```pageLoad```, and ```implicit``` though

**Comment by christian-bromann** [comment]
Closing in favor of #3218

**Pull request closed without merging**